### DISCUSSION TITLE
Implement generators

### DISCUSSION THREAD

**Discussion by Razican** [boa-dev/boa#1529] [Ideas]
**ECMASCript feature**
Currently, Boa has no support for Generators. Generators can be synchronous or asynchronous, and you can find the specs here:

 - Generator Function Definitions: https://tc39.es/ecma262/#sec-generator-function-definitions
 - Async Generator Function Definitions: https://tc39.es/ecma262/#sec-async-generator-function-definitions

This first needs changes in the parser, to check if a the function being defined is a generator or not. Will almost certainly require new AST nodes for that.

Then, it will require implementing the "Runtime Semantics" on the execution side. This will require implementing the new methods mentioned in the spec, and maybe some changes on how we handle #367. This might also collide with #818, since `async` generators will certainly require asynchronous execution that will be implemented there.

**Comment by raskad** [comment]
I took a look at how the spec implements the execution of generators. The best starting point to understand this is [here](https://tc39.es/ecma262/#sec-generatorstart).

Generators are implemented via switching the [execution context](https://tc39.es/ecma262/#sec-execution-contexts). We was already a discussion about those is https://github.com/boa-dev/boa/issues/948.
In my opinion it seems really difficult/hacky to implement execution contexts for our current tree walk interpreter. I would propose, that we block generator execution on the switch to the VM and implement execution contexts there.

**Reply by jedel1043** [reply]
+1, implementing async/generators is already pretty difficult, and having to do it twice would make us lose a lot of development time. Focusing our efforts on expanding the VM is, in my opinion, the best alternative.

**Reply by Razican** [reply]
Let's do it! Let's go full VM!

**Reply by jasonwilliams** [reply]
Yeah i hope to look at the VM this release

**Comment by jedel1043** [comment]
Closing since this has been implemented for a while now.
